platform/lib/routers/resourceFallbackHandler.js

```diff
@@ -0,0 +1,82 @@
+/**
+ * Copyright 2019 The AMPHTML Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+const fs = require('fs');
+const path = require('path');
+const express = require('express');
+const {setNoCache} = require('@lib/utils/cacheHelpers.js');
+
+const {Signale} = require('signale');
+
+const log = new Signale({
+  'interactive': false,
+  'scope': 'Resource Fallback Handler',
+});
+
+
+/**
+ *  Create a router to handle requests to files with hash that where not found.
+ *  It will serve the existing file with a different hash.
+ *
+ *  This will prevent a 404 error while CDN or cluster nodes have different versions of our app.
+ *  See https://github.com/ampproject/docs/issues/2029
+ */
+function createHandler(rootDir) {
+  // eslint-disable-next-line new-cap
+  const handler = express.Router();
+
+  /**
+   *  We assume that the hash in the URL is at least 15 characters long (webpack default is 20)
+   *  and has only digits and lowercase characters.
+   */
+  handler.get(/^(.*\/)([^/]+\.)[a-z0-9]{15,}(\.[^/]+)$/, (request, response, next) => {
+    const requestPath = request.params['0'];
+    const fileNamePrefix = request.params['1']; // including the dot at the end
+    const fileNamePostfix = request.params['2']; // starting with a dot
+
+    try {
+      const folder = path.join(rootDir, requestPath);
+      const availableFile = findFile(folder, fileNamePrefix, fileNamePostfix);
+
+      if (availableFile) {
+        setNoCache(response);
```

Comment:
Thinking about this again, it's probably OK to use a less strict caching policy (e.g. e-tag based caching would be OK).

Reply:
We do not send the correct content for the requested URL, but a newer or older version. So the response should not be cached, especially when we serve an older version.
What we do is not the ideal solution to the problem, but better than a 404 error.

Reply:
If the response content has not changed, the browser does not need to re-download it. So setting `max-age: 0` and `ETag: 123455` might be a better approach. Afaik express sets the Etag by default. `setNoCache`currently force the browser to always re-download.

https://developers.google.com/web/fundamentals/performance/optimizing-content-efficiency/http-caching#validating_cached_responses_with_etags

Reply:
When the content is not changed, webpack should not generate new hashes for the files.
So I would leave it at that, since it is an edge case anyway.

Reply:
I'm talking about the Etags generated by express based on the response content. Use case would be reloads while the js/css assets are stale.

Reply:
Anyway, no need to change anything. Browsers make use of Etags even if `private` and `no-cache` directives are given.